File: core/src/galenscovell/hinterstar/ui/components/gamescreen/hud/WeaponSelectPanel.scala
package galenscovell.hinterstar.ui.components.gamescreen.hud

import com.badlogic.gdx.scenes.scene2d.{InputEvent, Touchable}
import com.badlogic.gdx.scenes.scene2d.ui._
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.badlogic.gdx.utils.Align
import galenscovell.hinterstar.things.parts.Weapon
import galenscovell.hinterstar.things.ships.Ship
import galenscovell.hinterstar.util.{CrewOperations, Resources}


class WeaponSelectPanel(rootShip: Ship) extends Table {
  this.setFillParent(true)
  this.setTouchable(Touchable.enabled)



  def refresh(weapons: Array[Weapon], subsystem: String, crewmate: String): Unit = {
    this.clear()

    val selectionPanel: Table = new Table
    selectionPanel.setBackground(Resources.npDarkBlue)

    val topTable: Table = new Table
    topTable.setBackground(Resources.npDarkGray)
    val crewmateLabel: Label = new Label(crewmate, Resources.labelTinyStyle)
    crewmateLabel.setAlignment(Align.center)
    val closeButton: TextButton = new TextButton("Close", Resources.greenButtonStyle)
    closeButton.addListener(new ClickListener() {
      override def clicked(event: InputEvent, x: Float, y: Float): Unit = {
        CrewOperations.cancelWeaponAssignment()
      }
    })

    topTable.add(crewmateLabel).expand.fill.width(200).height(20).left
    topTable.add(closeButton).expand.fill.width(200).height(20).right
    selectionPanel.add(topTable).expand.fill.height(20).center

    selectionPanel.row

    val weaponTableHolder: Table = new Table

    for (weapon: Weapon <- weapons) {
      val weaponTable: Table = new Table

      if (weapon.getSubsystem == subsystem) {
        if (!weapon.isActive) {
          weaponTable.setBackground(Resources.npDarkGray)
        } else {
          weaponTable.setBackground(Resources.npGreen)
        }
      } else {
        weaponTable.setBackground(Resources.npGray)
      }

      val topBarTable: Table = new Table
      topBarTable.setBackground(Resources.npBlue)

      val iconTable = new Table
      iconTable.setBackground(Resources.npGreen)

      val damageLabel: Label = new Label(weapon.getDamage.toString, Resources.labelSmallStyle)
      damageLabel.setAlignment(Align.center)

      topBarTable.add(iconTable).expand.fill.width(20).height(20)
      topBarTable.add(damageLabel).expand.fill.width(80).height(20)

      val weaponImage: Table = new Table
      weaponImage.setTouchable(Touchable.enabled)

      if (!weapon.isActive && weapon.getSubsystem == subsystem) {
        weaponImage.addListener(new ClickListener() {
          override def clicked(event: InputEvent, x: Float, y: Float): Unit = {
            CrewOperations.selectWeapon(weapon)
          }
        })
      }

      val bottomWeaponTable: Table = new Table
      bottomWeaponTable.setBackground(Resources.npDarkBlue)
      val weaponLabel: Label = new Label(weapon.getName, Resources.labelSmallStyle)
      weaponLabel.setAlignment(Align.center)

      bottomWeaponTable.add(weaponLabel).expand.fill

      weaponTable.add(topBarTable).expand.fill.height(20)
      weaponTable.row
      weaponTable.add(weaponImage).expand.fill.height(50)
      weaponTable.row
      weaponTable.add(bottomWeaponTable).expand.fill.height(20)

      weaponTableHolder.add(weaponTable).expand.width(100).pad(5).left
    }

    selectionPanel.add(weaponTableHolder).expand.left

    this.add(selectionPanel).expand.fill.height(120).center.bottom
  }
}

File: core/src/galenscovell/hinterstar/ui/components/gamescreen/hud/WeaponSelectPanel.kt
package galenscovell.hinterstar.ui.components.gamescreen.hud

import com.badlogic.gdx.scenes.scene2d.*
import com.badlogic.gdx.scenes.scene2d.ui.*
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.badlogic.gdx.utils.Align
import galenscovell.hinterstar.things.parts.Weapon
import galenscovell.hinterstar.things.ships.Ship
import galenscovell.hinterstar.util.*


class WeaponSelectPanel(rootShip: Ship) : Table() {


    fun refresh(weapons: List<Weapon>, subsystem: String, crewmate: String): Unit {
        this.clear()
        this.setFillParent(true)
        this.touchable = Touchable.enabled

        val selectionPanel: Table = Table()
        selectionPanel.background = Resources.npTest1

        val topTable: Table = Table()
        topTable.background = Resources.npTest4
        val crewmateLabel: Label = Label(crewmate, Resources.labelTinyStyle)
        crewmateLabel.setAlignment(Align.center)
        val closeButton: TextButton = TextButton("Close", Resources.greenButtonStyle)
        closeButton.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent, x: Float, y: Float): Unit {
                CrewOperations.cancelWeaponAssignment()
            }
        })

        topTable.add(crewmateLabel).expand().fill().width(200f).height(20f).left()
        topTable.add(closeButton).expand().fill().width(200f).height(20f).right()
        selectionPanel.add(topTable).expand().fill().height(20f).center()

        selectionPanel.row()

        val weaponTableHolder: Table = Table()

        for (weapon: Weapon in weapons) {
            val weaponTable: Table = Table()

            if (weapon.getSubsystem() == subsystem) {
                if (!weapon.isActive()) {
                    weaponTable.background = Resources.npTest4
                } else {
                    weaponTable.background = Resources.npTest0
                }
            } else {
                weaponTable.background = Resources.npTest2
            }

            val topBarTable: Table = Table()
            topBarTable.background = Resources.npTest3

            val iconTable = Table()
            iconTable.background = Resources.npTest0

            val damageLabel: Label = Label(weapon.getDamage().toString(), Resources.labelSmallStyle)
            damageLabel.setAlignment(Align.center)

            topBarTable.add(iconTable).expand().fill().width(20f).height(20f)
            topBarTable.add(damageLabel).expand().fill().width(80f).height(20f)

            val weaponImage: Table = Table()
            weaponImage.touchable = Touchable.enabled

            if (!weapon.isActive() && weapon.getSubsystem() == subsystem) {
                weaponImage.addListener(object : ClickListener() {
                    override fun clicked(event: InputEvent, x: Float, y: Float): Unit {
                        CrewOperations.selectWeapon(weapon)
                    }
                })
            }

            val bottomWeaponTable: Table = Table()
            bottomWeaponTable.background = Resources.npTest1
            val weaponLabel: Label = Label(weapon.getName(), Resources.labelSmallStyle)
            weaponLabel.setAlignment(Align.center)

            bottomWeaponTable.add(weaponLabel).expand().fill()

            weaponTable.add(topBarTable).expand().fill().height(20f)
            weaponTable.row()
            weaponTable.add(weaponImage).expand().fill().height(50f)
            weaponTable.row()
            weaponTable.add(bottomWeaponTable).expand().fill().height(20f)

            weaponTableHolder.add(weaponTable).expand().width(100f).pad(5f).left()
        }

        selectionPanel.add(weaponTableHolder).expand().left()

        this.add(selectionPanel).expand().fill().height(120f).center().bottom()
    }
}
